pkg/controller/elasticsearch/driver/downscale_invariants.go

```diff
@@ -16,30 +16,37 @@ import (
 const (
 	OneMasterAtATimeInvariant        = "A master node is already in the process of being removed"
 	AtLeastOneRunningMasterInvariant = "Cannot remove the last running master node"
+	RespectMaxUnavailableInvariant   = "Not removing node to respect maxUnavailable setting"
 )
 
 // checkDownscaleInvariants returns true if the given state state allows downscaling the given StatefulSet.
 // If not, it also returns the reason why.
 func checkDownscaleInvariants(state downscaleState, statefulSet appsv1.StatefulSet) (bool, string) {
-	if !label.IsMasterNodeSet(statefulSet) {
-		// only care about master nodes
-		return true, ""
+	if state.removalsAllowed != nil && *state.removalsAllowed < 1 {
+		// removalsAllowed is nil then removals shouldn't be bounded here
+		return false, RespectMaxUnavailableInvariant
 	}
-	if state.masterRemovalInProgress {
-		return false, OneMasterAtATimeInvariant
-	}
-	if state.runningMasters == 1 {
-		return false, AtLeastOneRunningMasterInvariant
+
+	if label.IsMasterNodeSet(statefulSet) {
+		if state.masterRemovalInProgress {
+			return false, OneMasterAtATimeInvariant
+		}
+		if state.runningMasters == 1 {
+			return false, AtLeastOneRunningMasterInvariant
+		}
 	}
+
 	return true, ""
 }
 
 // downscaleState tracks the state of a downscale to be checked against invariants
 type downscaleState struct {
-	// masterRemovalInProgress indicates whether a master node is in the process of being removed already.
-	masterRemovalInProgress bool
 	// runningMasters indicates how many masters are currently running in the cluster.
 	runningMasters int
+	// removalsAllowed indicates how many nodes can be removed to adhere to maxUnavailable setting
+	removalsAllowed *int32
```

Comment:
Can this be `nil`? When/how? I'm starting to wonder whether we should make it possible to be `nil` (unbounded), and how the user would express that.

Reply:
After discussing it with David: it can indeed be `Nil`, if the user specifies a negative number in the spec.
So we have the CRD schema:

* nil means "default value"
* positive means well positive :)
* negative (-1) means unbounded

And the code schema:
* nil means unbounded
* positive means positive
* negative is not possible

Which sounds good to me (maybe worth a few more comments to make it explicit).